New stdlib function array.group_by (#3710)
grafana/alloy: docs/sources/reference/stdlib/array.md

[tests]: https://github.com/grafana/alloy/blob/main/syntax/vm/vm_stdlib_test.go
[experimental]: https://grafana.com/docs/release-life-cycle/

## array.group_by

{{< docs/shared lookup="stability/experimental_feature.md" source="alloy" version="<ALLOY_VERSION>" >}}

The `array.group_by` function groups an array of objects by a given key.

* The first argument is an array of objects.
* The second argument is a string that is the key to group by.
* The third argument is a boolean that indicates whether the elements that don't match the key should be dropped (true) or added to the empty group (false).

### Examples

```alloy

> > array.group_by([{"name" = "alice", "age" = "20"}, {"name" = "bob", "age" = "20"}, {"name" = "charlie", "age" = "30"}, {"name" = "william"}], "age", false)

Author: @clayton-cornell the linting is a bit broken in this example because of the "". Do you know why?

Reviewer: I see there's been changes. I guess you've figured it out?

Author: no there is something off but it's in general. If you look at the examples on this page: https://grafana.com/docs/alloy/latest/reference/stdlib/array/#examples-1 like the first array.combine_maps([{"instance"="1.1.1.1", "team"="A"}], [{"instance"="1.1.1.1", "cluster"="prod"}], ["instance"])
[{"instance"="1.1.1.1", "team"="A", "cluster"="prod"}]  the colors are off

Author: I merged it for now, we can fix it in another PR

[{"age" = "20", "items" = [{"name" = "alice", "age" = "20"}, {"name" = "bob", "age" = "20"}]}, {"age" = "30", "items" = [{"name" = "charlie", "age" = "30"}]}, {"age" = "", "items" = [{"name" = "william"}]}]

> array.group_by([{"name" = "alice", "age" = "20"}, {"name" = "bob", "age" = "20"}, {"name" = "charlie", "age" = "30"}, {"name" = "william"}], "age", true)
[{"age" = "20", "items" = [{"name" = "alice", "age" = "20"}, {"name" = "bob", "age" = "20"}]}, {"age" = "30", "items" = [{"name" = "charlie", "age" = "30"}]}]
```

The following example shows how to use `array.group_by` function with a `foreach` block to group targets by match labels and create a `prometheus.scrape` component for each group dynamically. The targets in this example are expected to have a label "match" that contains instant vector selectors separated by slash (see [Federation][federation]).

```alloy
foreach "federation" {
 collection = array.group_by(discovery.file.example.targets, "match", false)
 var = "each"

>  id  = "match"

Author: the id can be long sometimes which is not super ideal, I'm thinking about adding the possibility to hash it via an additional arg

> The following example shows how to use `array.group_by` function with a `foreach` block to group targets by match labels and create a `prometheus.scrape` component for each group dynamically. The targets in this example are expected to have a label "match" that contains instant vector selectors separated by slash (see [Federation][federation] for more information on the match parameter).

Reviewer: ```suggestion
The following example shows how to use the `array.group_by` function with a `foreach` block to group targets by match labels and create a `prometheus.scrape` component for each group dynamically.
The targets in this example should have a label "match" that contains instant vector selectors separated by slash (refer to [Federation][federation] for more information on the match parameter).
```
